File: _backup/v36-pre-navegacao-2026-07-13/test_a11y.py
"""Teste automatizado — home + menu acessibilidade (Playwright).
Servidor interno na porta 8095 — ver README (dev manual: porta 8093)."""
import json
import sys
import time
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from threading import Thread

ROOT = Path(__file__).resolve().parent
PORT = 8095


def start_server():
    class Handler(SimpleHTTPRequestHandler):
        def __init__(self, *args, **kwargs):
            super().__init__(*args, directory=str(ROOT), **kwargs)

    server = ThreadingHTTPServer(("127.0.0.1", PORT), Handler)
    Thread(target=server.serve_forever, daemon=True).start()
    return server


def metrics(page):
    return page.evaluate(
        """() => {
            const app = document.querySelector('.app');
            const home = document.getElementById('screenHome');
            const active = document.querySelector('.screen.active');
            const cs = app ? getComputedStyle(app) : null;
            return {
                appH: app ? app.offsetHeight : 0,
                homeDisplay: home ? getComputedStyle(home).display : null,
                activeId: active ? active.id : null,
                appOpacity: cs ? cs.opacity : null,
                appFilter: cs ? cs.filter : 'none',
                bodyClass: document.body ? document.body.className : '',
            };
        }"""
    )


def check(page, label, results, errors):
    m = metrics(page)
    ok = m["appH"] > 200 and m["activeId"] == "screenHome" and m["homeDisplay"] == "block"
    results.append({"label": label, "ok": ok, **m})
    if not ok:
        raise RuntimeError(f"{label} FALHOU: {m}")
    if errors:
        print(f"  avisos em {label}:", errors[:3])


def main():
    try:
        from playwright.sync_api import sync_playwright
    except ImportError:
        print("Instalando playwright...")
        import subprocess

        subprocess.check_call([sys.executable, "-m", "pip", "install", "playwright", "-q"])
        subprocess.check_call([sys.executable, "-m", "playwright", "install", "chromium"])
        from playwright.sync_api import sync_playwright

    server = start_server()
    results = []
    errors = []
    url = f"http://127.0.0.1:{PORT}/index.html?v=testpy"

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()
        page.on("pageerror", lambda e: errors.append(str(e)))
        page.on("console", lambda msg: errors.append(msg.text) if msg.type == "error" else None)

        # 1) Carregamento limpo
        page.goto(url, wait_until="networkidle")
        time.sleep(0.6)
        check(page, "carregamento inicial", results, errors)
        errors.clear()

        # 2) Estado salvo problemático
        page.evaluate(
            """() => localStorage.setItem('guardiao_vale_v7', JSON.stringify({
                lang:'pt', onboardingDone:true,
                a11y:{voice:false,contrast:true,large:false,motion:true,signs:true,
                      fontScale:0,links:false,spacing:false,letterSpace:false,
                      dyslexia:false,colorblind:'protanopia',readingMode:false}
            }))"""
        )
        page.reload(wait_until="networkidle")
        time.sleep(1.2)
        check(page, "estado salvo corrompido", results, errors)
        errors.clear()

        # 3) Menu A11y — todas as opções
        page.click("#a11yBtn")
        page.wait_for_selector("#a11yMenu:not([hidden])", timeout=5000)
        toggles = page.locator("#a11yMenu .am-toggle")
        n = toggles.count()
        for i in range(n):
            toggles.nth(i).click()
            time.sleep(0.25)
            check(page, f"toggle {i+1}/{n}", results, errors)
            errors.clear()

        # 4) Restaurar padrão
        page.click("#resetA11yMenuBtn")
        time.sleep(0.4)
        check(page, "restaurar padrão", results, errors)

        # 5) Toggles da home
        for opt in ["optVoice", "optContrast", "optSigns", "optMotion"]:
            page.locator(f"#{opt}").click(force=True)
            time.sleep(0.2)
            check(page, f"home {opt} on", results, errors)
            errors.clear()
            page.locator(f"#{opt}").click(force=True)
            time.sleep(0.15)

        browser.close()

    server.shutdown()
    print("\n[OK] TODOS OS TESTES PASSARAM\n")
    for r in results:
        print(f"  OK {r['label']} — appH={r['appH']} filter={r.get('appFilter','none')}")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        print("\n[FALHA] TESTE FALHOU\n", exc)
        raise SystemExit(1)

File: test-a11y.mjs
/**
 * Teste automatizado — página inicial + menu de acessibilidade
 * Uso: node test-a11y.mjs
 * Servidor interno na porta 8094 — ver README (dev manual: porta 8093).
 */
import { chromium } from "playwright";
import { createServer } from "http";
import { readFileSync, statSync } from "fs";
import { join, extname } from "path";
import { fileURLToPath } from "url";

const ROOT = fileURLToPath(new URL(".", import.meta.url));
const PORT = 8094;

const MIME = {
  ".html": "text/html",
  ".css": "text/css",
  ".js": "application/javascript",
  ".svg": "image/svg+xml",
  ".png": "image/png",
};

function startServer() {
  return new Promise((resolve) => {
    const server = createServer((req, res) => {
      const url = req.url.split("?")[0] || "/";
      const file = join(ROOT, url === "/" ? "index.html" : url.replace(/^\//, ""));
      try {
        const data = readFileSync(file);
        res.writeHead(200, { "Content-Type": MIME[extname(file)] || "application/octet-stream" });
        res.end(data);
      } catch {
        res.writeHead(404).end("Not found");
      }
    });
    server.listen(PORT, () => resolve(server));
  });
}

async function assertVisible(page, selector, label) {
  const el = page.locator(selector).first();
  await el.waitFor({ state: "attached", timeout: 8000 });
  const box = await el.boundingBox();
  if (!box || box.width < 2 || box.height < 2) {
    throw new Error(`${label}: elemento sem área visível (${selector})`);
  }
}

async function appMetrics(page) {
  return page.evaluate(() => {
    const app = document.querySelector(".app");
    const home = document.getElementById("screenHome");
    const active = document.querySelector(".screen.active");
    const style = app ? getComputedStyle(app) : null;
    return {
      appH: app ? app.offsetHeight : 0,
      homeDisplay: home ? getComputedStyle(home).display : null,
      activeId: active ? active.id : null,
      appOpacity: style ? style.opacity : null,
      appFilter: style ? style.filter : null,
      bodyClass: document.body.className,
      errors: window.__testErrors || [],
    };
  });
}

async function run() {
  const server = await startServer();
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext();
  const page = await context.newPage();

  const errors = [];
  page.on("pageerror", (e) => errors.push(String(e)));
  page.on("console", (msg) => {
    if (msg.type() === "error") errors.push(msg.text());
  });

  await page.addInitScript(() => {
    window.__testErrors = [];
    window.addEventListener("error", (e) => window.__testErrors.push(e.message));
  });

  const url = `http://127.0.0.1:${PORT}/index.html?v=test`;
  const results = [];

  async function check(label) {
    const m = await appMetrics(page);
    const ok = m.appH > 200 && m.activeId === "screenHome" && m.homeDisplay === "block";
    results.push({ label, ok, ...m, errors: [...errors] });
    if (!ok) throw new Error(`${label} FALHOU: appH=${m.appH} active=${m.activeId}`);
  }

  try {
    // 1) Carregamento limpo
    await page.goto(url, { waitUntil: "networkidle" });
    await page.waitForTimeout(600);
    await assertVisible(page, "#screenHome", "Home");
    await check("carregamento inicial");
    results[results.length - 1].errors = [];

    // 2) Simular estado corrompido (daltônico + libras + contraste)
    await page.evaluate(() => {
      localStorage.setItem(
        "guardiao_orbita_v7",
        JSON.stringify({
          lang: "pt",
          a11y: {
            voice: false,
            contrast: true,
            large: false,
            motion: true,
            signs: true,
            fontScale: 0,
            links: false,
            spacing: false,
            letterSpace: false,
            dyslexia: false,
            colorblind: "protanopia",
            readingMode: false,
          },
        })
      );
    });
    await page.reload({ waitUntil: "networkidle" });
    await page.waitForTimeout(1200);
    await check("estado salvo corrompido");
    results[results.length - 1].errors = [];

    // 3) Abrir menu A11y e clicar em todas as opções
    await page.click("#a11yBtn");
    await page.waitForSelector("#a11yMenu:not([hidden])", { timeout: 3000 });
    const toggles = page.locator("#a11yMenu .am-toggle");
    const count = await toggles.count();
    for (let i = 0; i < count; i++) {
      await toggles.nth(i).click();
      await page.waitForTimeout(200);
      await check(`toggle ${i + 1}/${count}`);
      results[results.length - 1].errors = [];
    }

    // 4) Restaurar padrão
    await page.click("#resetA11yMenuBtn");
    await page.waitForTimeout(400);
    await check("restaurar padrão");

    // 5) Toggles da home
    for (const id of ["optVoice", "optContrast", "optSigns", "optMotion"]) {
      await page.locator(`#${id}`).click({ force: true });
      await page.waitForTimeout(200);
      await check(`home ${id}`);
      results[results.length - 1].errors = [];
      await page.locator(`#${id}`).click({ force: true });
      await page.waitForTimeout(150);
    }

    console.log("\n✅ TODOS OS TESTES PASSARAM\n");
    results.forEach((r) => {
      console.log(`  ✓ ${r.label} — appH:${r.appH} filter:${r.appFilter || "none"}`);
    });
    process.exit(0);
  } catch (err) {
    console.error("\n❌ TESTE FALHOU\n", err.message);
    results.forEach((r) => {
      console.log(`  ${r.ok ? "✓" : "✗"} ${r.label}`, JSON.stringify(r));
    });
    if (errors.length) console.log("Console errors:", errors);
    process.exit(1);
  } finally {
    await browser.close();
    server.close();
  }
}

run();
